js/copy-code.js

```diff
@@ -0,0 +1,88 @@
+const COPY_CODE_CLASS = "usa-copy-code";
+const COPY_BUTTON_CLASS = `usa-button usa-button--outline ${COPY_CODE_CLASS}__button`;
+const COPY_BUTTON_SUCCESS_CLASS = `${COPY_CODE_CLASS}__button--success`;
+const COPY_WRAPPER_CLASS = `${COPY_CODE_CLASS}__wrapper`;
+
+const COPY_CODE = document.querySelectorAll(`.${COPY_CODE_CLASS}`);
+
+/**
+ * Creates a wrapper for copy button.
+ *
+ * @return {HTMLElement} - A div for `usa-copy-code__wrapper`.
+ */
+const createWrapper = () => {
+  const wrapper = document.createElement("div");
+  wrapper.classList.add(COPY_WRAPPER_CLASS);
+
+  return wrapper;
+};
+
+/**
+ * Creates a copy button component.
+ *
+ * @return {HTMLElement} - A button for copying code `usa-copy-code__button`.
+ */
+const createCopyButton = () => {
+  const btn = document.createElement("button");
+  const btnText = `
+    <span class="" aria-hidden="true">Copy</span>
+    <span class="usa-sr-only">Copy component code</span>
+  `;
+
+  btn.className = COPY_BUTTON_CLASS;
+  btn.setAttribute("type", "button");
+
+  btn.insertAdjacentHTML("beforeend", btnText);
+
+  return btn;
+};
+
+/**
+ * Copy <code> text content when copy code button is clicked.
+ *
+ * @param {Event} event - The click event from copy code button.
+ * @return {Clipboard} - Code element contents written to clipboard.
+ */
+const copyOnClick = (event) => {
+  const copyBtn = event.currentTarget;
+  const labelVisual = copyBtn.querySelector("[aria-hidden]");
+  const labelSROnly = copyBtn.querySelector(".usa-sr-only");
+
+  // Set success state
+  copyBtn.classList.add(COPY_BUTTON_SUCCESS_CLASS);
+  labelVisual.textContent = "Copied!";
+  labelSROnly.textContent = "Code copied to clipboard";
+
+  // After timeout, reset to default state
+  setTimeout(() => {
+    copyBtn.classList.remove(COPY_BUTTON_SUCCESS_CLASS);
+    labelVisual.textContent = "Copy";
+    labelSROnly.textContent = "Copy component code";
+  }, 3000);
```

Comment:
I noticed that in VoiceOver, if you keep focus on the button after the copy action, the status readout goes from “Code copied to clipboard” to “copy component code” after a few seconds. I wonder if this could lead to confusion for screen reader users. Can it stay in the success state for a longer period? Maybe even something like 30s or 60s?

Not sure if action must be taken here, but wanted to flag for consideration.

Reply:
I wonder if the better pattern here would be to _not_ read out that the button has reset? Instead, only have it read the button again once it has been focused again. This is similar to how GitHub handles their read out.

```
Try copying me with your screen reader!

1. First hover with mouse
2. Use arrow key to select button
3. Reads "copy"
4. Reads "copied"
5. Does not read "copy" after button is reset
```

@amycole501 Any thoughts on having screen readers not read the state change from "success/copied" to default?

Reply:
In JAWS nothing is announced once the code is copied. It's silent and the person must assume it worked.
Same behavior (or lack thereof) in NVDA and narrator.

What would make the most sense is having the announcement "Code copied to clipboard". Then I don't think the person would care if the state goes back to the original or not and wouldn't need to hear it. No need to declare it's back to the original state unless the person has to scroll up and start over for some reason.

Reply:
@amycole501  just curious, what browser did you test in?

Reply:
I would love to have the button not read out once it returns to the default state. I think it's unnecessary for the user and they realistically only need to hear "copy component code" when the button receives focus.

I've tried the following with inconsistent results:

### Separate Span for its initial selection and for the success text:

Something along the lines of:

```html
<div class="copy-code__wrapper">
  <div class="usa-sr-only copy-code__sr-status" id="range-slider-code-content__sr-label" role="status">
      Copy component code.
  </div>
  <button class="usa-button usa-button--outline copy-code__button copy-code__button--success" type="button" aria-labelledby="range-slider-code-content__sr-label">
      <span aria-hidden="true">Copied!</span>
      <span aria-live="true">Code copied to clipboard</span>
   </button>
</div>
```

This works well at first, but after the button is triggered once there is a delay until it will read out the success state again (in chrome). This _might_ not be an issue in practice if the user doesn't attempt to copy the same code multiple times at once or within a few seconds.

If I activate the button and wait a few seconds, or navigate around the page before returning back and activating the button again, it will read as intended!

### Remove aria tags until after the success state is removed

I was hoping that if we add `aria-live/aria-labelledby` _after_ the button returns to it's default state, it would not read out the default message again. Unfortunately, the screen reader will read out the message the moment the aria tags have returned to the button

### Add the default message on focus

I'm trying various ways of only adding the default message once the button is focused, that way we don't need to update the text content when it returns to its default state. I haven't gotten it working as I'd like yet but I want to tinker with it a bit more before I throw in the towel on this method

Reply:
Resolved in https://github.com/uswds/uswds-site/pull/2059/commits/e032137ef1776b1437a5cbfc484a35e0ee279e89